File: 2024/Day 01/Python/main.py
"""Advent of Code - 2024 - Day 1"""
from collections import Counter

with open(0, encoding="utf-8") as f:
    lists = [line.split("   ") for line in f.read().splitlines()]
left = [int(line[0]) for line in lists]
right = [int(line[1]) for line in lists]

print(sum(abs(l-r) for (l, r) in zip(sorted(left), sorted(right))))

c = Counter(right)
print(sum(l*c[l] for l in left))

File: 2024/Day 01/Rust/src/main.rs
use itertools::Itertools;
use std::io;

fn main() {
    let lines = io::stdin()
        .lines()
        .collect::<Result<Vec<String>, io::Error>>()
        .unwrap();
    let rows: Vec<Vec<&str>> = lines
        .iter()
        .map(|x| x.split("   ").collect::<Vec<_>>())
        .collect();
    let mut left: Vec<i32> = Vec::new();
    let mut right: Vec<i32> = Vec::new();
    for row in rows {
        left.push(row[0].parse().unwrap());
        right.push(row[1].parse().unwrap());
    }
    left.sort();
    right.sort();
    let result1: i32 = left
        .iter()
        .zip(right.iter())
        .map(|(l, r)| i32::abs(l - r))
        .sum();
    println!("{}", result1);
    let counter = right.into_iter().counts();
    let result2: usize = left
        .into_iter()
        .map(|x| x as usize * counter.get(&x).unwrap_or(&0))
        .sum();
    println!("{}", result2);
}
